client/consensus/common/src/level_monitor.rs

```diff
@@ -0,0 +1,269 @@
+where
+	Block: BlockT,
+	BE: Backend<Block>,
+{
+	/// Restore the structure using the backend.
+	///
+	/// Blocks freshness values are inferred from the height and not from the effective import
+	/// moment. This is a not accurate but "good-enough" best effort solution.
+	///
+	/// Level limits are not enforced during this phase.
+	fn restore(&mut self) {
+		let mut counter_max = 0;
+		let finalized = self.backend.blockchain().info().finalized_hash;
+
+		for leaf in self.backend.blockchain().leaves().unwrap_or_default() {
+			let route = sp_blockchain::tree_route(self.backend.blockchain(), finalized, leaf)
+				.expect("Route from finalized to leaf should be available; qed");
+			if !route.retracted().is_empty() {
+				continue
+			}
+			std::iter::once(route.common_block()).chain(route.enacted()).for_each(|elem| {
+				if !self.freshness.contains_key(&elem.hash) {
+					// Use the block height value as the freshness.
+					self.import_counter = elem.number.unique_saturated_into();
+					self.update(elem.number, elem.hash);
+				}
+			});
+			if self.import_counter > counter_max {
+				counter_max = self.import_counter;
+			}
+		}
+
+		self.import_counter = counter_max;
+		self.lowest_level = self.backend.blockchain().info().finalized_number;
+	}
+
+	/// Check and enforce the limit bound at the given height.
+	///
+	/// In practice this will enforce the given height in having a number of blocks less than
+	/// the limit passed to the constructor.
+	///
+	/// If the given level is found to have a number of blocks greater than or equal the limit
+	/// then the limit is enforced by chosing one (or more) blocks to remove.
+	/// The removal strategy is driven by the block freshness.
+	///
+	/// A block freshness is determined by the most recent leaf freshness descending from the block
+	/// itself. In other words its freshness is equal to its more "fresh" descendant.
+	///
+	/// The least "fresh" blocks are eventually removed.
+	pub fn enforce_limit(&mut self, number: NumberFor<Block>) {
+		if self.import_counter == 0 {
+			// First call detected, restore using the backend persistent information.
+			self.restore();
+		}
+
+		let level = self.levels.entry(number).or_default();
+		if level.len() < self.level_limit {
+			return
+		}
+
+		log::debug!(
+			target: "parachain",
+			"Detected leaves overflow at height {}, removing obsolete blocks", number
+		);
+
+		let mut leaves = self.backend.blockchain().leaves().unwrap_or_default();
+		// Sort leaves by freshness (most recent first).
+		leaves.sort_unstable_by(|a, b| self.freshness.get(a).cmp(&self.freshness.get(b)));
+
+		// This may not be the most efficient way to remove **multiple** entries, but is the clear
+		// one. Should be considered that in normal conditions the number of blocks to remove is 0
+		// or 1, it is not worth to complicate the code too much.
+		let remove_count = level.len() - self.level_limit + 1;
+		(0..remove_count).for_each(|_| self.remove_block(number, &leaves));
```

Comment:
Removing one single block is the "normal" case.

I've added this because it may happen that we require to delete more than one block in the following cases:
1. the limit is changed  to something smaller. In this case the DB was previously pruned using the old bigger limit.
   The first time this code runs it will remove more than one block to enforce the limit.
2. The first time this code is enacted. This is a special case of 1 (with previous "bigger" limit equal to infinite).